Add [dynamic-yaml] badge

jacobtomlinson opened the pull request:
This PR closes #1617 by adding support for a dynamic YAML badge.

I have added `js-yaml` to handle the YAML parsing, but then the rest of the work has mostly been duplicating the JSON dynamic badge as YAML is just a superset of JSON.

I've tested it locally and run the test suite with no issues showing in the tests I added.

Please let me know if there is anything you want me to change.

paulmelnikow commented on services/yaml/yaml.tester.js:
```diff
@@ -0,0 +1,70 @@
+
+t.create('YAML from url')
+  .get('.json?url=https://github.com/badges/shields/raw/master/package.json&query=$.name&style=_shields_test')
+  .expectJSON({ name: 'custom badge', value: 'gh-badges', colorB: colorsB.brightgreen });
```
Wouldn't these tests need to pull from a yml file instead of a json file?

paulmelnikow commented on frontend/components/usage.js:
```diff
@@ -123,6 +123,9 @@ export default class Usage extends React.PureComponent {
         </p>
+        <p>
+          <code>/badge/dynamic/yaml.svg?url=&lt;URL&gt;&amp;label=&lt;LABEL&gt;&amp;query=&lt;<a href="https://www.npmjs.com/package/jsonpath" target="_BLANK" title="JSONdata syntax">$.DATA.SUBDATA</a>&gt;&amp;colorB=&lt;COLOR&gt;&amp;prefix=&lt;PREFIX&gt;&amp;suffix=&lt;SUFFIX&gt;</code>
+        </p>
```
Could you also add `yaml` to the list [in the dynamic badge maker](https://github.com/badges/shields/blob/master/frontend/components/dynamic-badge-maker.js#L44)?

jacobtomlinson replied:
Done!

jacobtomlinson commented:
I thought it was valid to leave the JSON tests as JSON is valid YAML. I can definitely add more which explicitly tests the YAML syntax.

jacobtomlinson commented:
I have updated the tests to use a YAML data source and added `yaml` to the list of options in the dynamic badge maker as requested.

Please let me know if there is anything else you need me to change.

paulmelnikow commented:
JSON is valid YAML?! Mind = blown 😀 

Thanks for the updates. I'll give this another read, and perhaps someone else would like to as well.

platan commented on services/yaml/yaml.tester.js:
```diff
@@ -0,0 +1,70 @@
+  .expectJSON({ name: 'Package Name', value: 'no url specified', colorB: colorsB.red });
+
+t.create('No query specified')
+  .get('.json?url=https://raw.githubusercontent.com/kubernetes/charts/568291d6e476c39ca8322c30c3f601d0383d4760/stable/coredns/Chart.yaml&label=Package Name&style=_shields_test')
```
You've used a link to specified commit, which great in my opinion :-)

jacobtomlinson replied:
👍 I thought it would be consistent when testing.

platan commented on services/yaml/yaml.tester.js:
```diff
@@ -0,0 +1,70 @@
+module.exports = t;
+
+t.create('Connection error')
+  .get('.json?url=https://github.com/badges/shields/raw/master/package.json&query=$.name&label=Package Name&style=_shields_test')
```
Could you, just for the sake of a consistency, change an uri value to a yaml resource (from examples below)?

jacobtomlinson replied:
Whoops must've missed that one.

platan commented on services/yaml/yaml.tester.js:
```diff
@@ -0,0 +1,70 @@
+  .get('.json?url=https://raw.githubusercontent.com/kubernetes/charts/568291d6e476c39ca8322c30c3f601d0383d4760/stable/coredns/Chart.yaml&query=$.version')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'custom badge',
+    value: Joi.string().regex(/^\d+(\.\d+)?(\.\d+)?$/)
```
I think you can use an `isSemver` validator here: https://github.com/badges/shields/blob/66c5d914c73f39acd6bf44e3d69dffa2c17a20c8/services/xml/xml.tester.js#L55-L57 
Or even better you can change expect to 
```js
.expectJSON({ name: 'custom badge', value: '0.8.0' });
```
since we are using resource from a specific commit.

jacobtomlinson replied:
I just copied the test from the json dynamic badge. Happy to update if you prefer.

platan commented on frontend/components/dynamic-badge-maker.js:
```diff
@@ -42,6 +42,7 @@ export default class DynamicBadgeMaker extends React.Component {
           onChange={event => this.setState({ datatype: event.target.value })}>
           <option value="" disabled>data type</option>
           <option value="json">json</option>
+          <option value="yaml">yaml</option>
```
Can we sort them in an alphabetical order (json, xml, yaml)?

jacobtomlinson replied:
See other comment about order.

platan commented on services/yaml/yaml.tester.js:
```diff
@@ -0,0 +1,70 @@
+
+t.create('YAML from url | with prefix & suffix & label')
+  .get('.json?url=https://raw.githubusercontent.com/kubernetes/charts/568291d6e476c39ca8322c30c3f601d0383d4760/stable/coredns/Chart.yaml&query=$.version&prefix=v&suffix= dev&label=Shields')
+  .expectJSONTypes(Joi.object().keys({
```
What do you thin about using this approach?
```js
.expectJSON({ name: 'Shields', value: 'v0.8.0 dev' });
```

jacobtomlinson replied:
Much neater, again I just copied the other test.

platan commented on frontend/components/usage.js:
```diff
@@ -123,6 +123,9 @@ export default class Usage extends React.PureComponent {
           <code>/badge/dynamic/json.svg?url=&lt;URL&gt;&amp;label=&lt;LABEL&gt;&amp;query=&lt;<a href="https://www.npmjs.com/package/jsonpath" target="_BLANK" title="JSONdata syntax">$.DATA.SUBDATA</a>&gt;&amp;colorB=&lt;COLOR&gt;&amp;prefix=&lt;PREFIX&gt;&amp;suffix=&lt;SUFFIX&gt;</code>
         </p>
+        <p>
+          <code>/badge/dynamic/yaml.svg?url=&lt;URL&gt;&amp;label=&lt;LABEL&gt;&amp;query=&lt;<a href="https://www.npmjs.com/package/jsonpath" target="_BLANK" title="JSONdata syntax">$.DATA.SUBDATA</a>&gt;&amp;colorB=&lt;COLOR&gt;&amp;prefix=&lt;PREFIX&gt;&amp;suffix=&lt;SUFFIX&gt;</code>
```
Can we sort them in an alphabetical order (json, xml, yaml)?

jacobtomlinson replied:
Yes sure. I went with this order because the usage of json and yaml are the same and so thought I would keep them together.

jacobtomlinson commented:
Thanks for the review comments @platan. I'll get this PR updated later on with the request changes.

jacobtomlinson commented:
@platan I have made the changes!

platan commented:
@jacobtomlinson thanks for fixes. That was quick! The code looks great in my opinion. 

I have one extra thing. What would we like to see as a badge value when resource is not a valid yaml file? Something simple like `invalid` (https://github.com/badges/shields/blob/master/doc/service-tests.md#tutorial)? Currently we have such result for https://www.google.com/robots.txt
YAML: 
![image](https://user-images.githubusercontent.com/1526000/38145437-9a32d2d0-3449-11e8-94cf-1eee642865a6.png)
JSON:
![image](https://user-images.githubusercontent.com/1526000/38145474-d16e5954-3449-11e8-9d7b-ae7db39ef857.png)
And for https://img.shields.io/badge/x-y-z.png
YAML:
![image](https://user-images.githubusercontent.com/1526000/38145525-050e45da-344a-11e8-8027-5db7726ec227.png)
JSON:
![image](https://user-images.githubusercontent.com/1526000/38145515-fd1455ae-3449-11e8-81f9-7a5b14c3d2be.png)
XML badge returns 
![image](https://user-images.githubusercontent.com/1526000/38145542-18cf7de6-344a-11e8-865d-81b4f891ba48.png) for both resources. 

@paulmelnikow @RedSparr0w What do you think about this?

jacobtomlinson commented:
A consistent error message would be good. I think `invalid` would be fine but not sure if it accurately describes what the problem is. Perhaps something like `invalid data source`, `parse error` or `load error` would be more useful when troubleshooting.

Happy to go with whatever you think. Do you want me to implement it for all three dynamic badges in this PR or raise a new PR for it?

platan commented:
@jacobtomlinson I agree with you, `invalid data source` or `parse error` is better than `invalid`. In my opinion we can merge this PR and prepare a separate one for handling parsing errors. Let's wait for opinion of the rest of the team.

RedSparr0w commented:
Sounds good to me, I originally left it as the default error message so the user could try and fix the badge/problem themselves, but I think something like 'invalid json', 'invalid source', 'parse error' would work better

paulmelnikow commented:
My preference would be **invalid json** / **invalid xml** / **invalid yaml** which is terse and consistent with other badges.

platan commented:
@RedSparr0w or @paulmelnikow can you approve this PR? Then we could merge it and new PR with a better error message can be raised.

paulmelnikow commented on frontend/lib/badge-url.spec.js:
```diff
@@ -59,6 +59,28 @@ describe('Badge URL functions', function() {
+  test(dynamicBadgeUrl, () => {
+    const yamlUrl = 'http://example.com/foo.yaml';
+    const query = '$.bar';
+    const prefix = 'value: ';
```
Aren't these three tests of `dynamicBadgeUrl` testing exactly the same thing? Maybe we should strip out two of them.

platan replied:
Good point. Let's remove two of them.

paulmelnikow commented on server.js:
```diff
@@ -7503,6 +7504,13 @@ cache({
+      case 'yaml':
+        requestOptions = {
+          headers: {
+            Accept: 'application/yaml, text/yaml, text/plain'
```
From [this Quora answer](https://www.quora.com/What-is-the-correct-MIME-type-for-YAML-documents/answer/Charles-Banas) it seems we probably should use `'application/x-yaml, text/x-yaml, text/plain'` here. Unless someone has a better source saying something else?

platan replied:
github.com, search in code:
- `"text/x-yaml" content type` 39,389 code results
- `"text/yaml" content type` 79,485 code results 
- `"application/x-yaml" content type` 34,264 code results 
- `"application/yaml" content type` 21,948 code results 

Should we add all of them since it's not standardized?

jacobtomlinson replied:
Yes that's probably best.

jacobtomlinson commented:
I have removed the extra tests and added the MIME types as suggested by @paulmelnikow.

I have also merged in the latest master and run `npm install` again to generate a new `package-lock.json` which resolves the merge conflicts it was complaining about.

Should be good to merge now!

platan commented:
@jacobtomlinson thanks. Code looks OK. I have one doubt about package-lock.json. In all dependencies of fsevents "bundled" (https://docs.npmjs.com/files/package-lock.json#bundled) property is replaced with "resolved" and "integrity". I'm not sure if this change changes anything. 4 months ago we had change from "resolved"+"integrity" to "bundled" https://github.com/badges/shields/commit/0da212b7deb95c0239e8c24899c846a9a2d9c038. 
@jacobtomlinson @paulmelnikow What do you think about this? Maybe this depends on a NPM version? Should we stick to a specific version of NPM?

jacobtomlinson commented:
I don't feel I can add useful input here. To be honest this is the first
project I've ever seen which tracks the package lock, I've always seen it
in the `.gitignore`.

Happy to change it however you wish.

On Tue, 3 Apr 2018, 17:53 Marcin Mielnicki, <notifications@github.com>
wrote:

> @jacobtomlinson <https://github.com/jacobtomlinson> thanks. Code looks
> OK. I have one doubt about package-lock.json. In all dependencies of
> fsevents "bundled" (https://docs.npmjs.com/files/package-lock.json#bundled)
> property is replaced with "resolved" and "integrity". I'm not sure if this
> change changes anything. 4 months ago we had change from
> "resolved"+"integrity" to "bundled" 0da212b
> <https://github.com/badges/shields/commit/0da212b7deb95c0239e8c24899c846a9a2d9c038>
> .
> @jacobtomlinson <https://github.com/jacobtomlinson> @paulmelnikow
> <https://github.com/paulmelnikow> What do you think about this? Maybe
> this depends on a NPM version? Should we stick to a specific version of NPM?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/badges/shields/pull/1623#issuecomment-378320114>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABiUYq4RjWML5vkHwpA3O8aBzk_nhVLnks5tk6kbgaJpZM4TAk-F>
> .
>

platan commented:
@jacobtomlinson Thanks for your contribution and your patience 👍 Would  you like to prepare a next PR with an error message when data source is invalid? You could try with messages proposed by Paul (https://github.com/badges/shields/pull/1623#issuecomment-377796115).    

@RedSparr0w Thank you for a review.
